File: June/29-06-2024/Identical Linked Lists.js
//{ Driver Code Starts
// Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString =
        inputString.trim().split('\n').map(string => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {
        let n1 = parseInt(readLine());
        let input_ar1 = readLine().split(' ').map(x => parseInt(x));
        let head1 = new Node(input_ar1[0]);
        let tail1 = head1;
        for (let i = 1; i < n1; i++) {
            tail1.next = new Node(input_ar1[i]);
            tail1 = tail1.next;
        }
        let n2 = parseInt(readLine());
        let input_ar2 = readLine().split(' ').map(x => parseInt(x));
        let head2 = new Node(input_ar2[0]);
        let tail2 = head2;
        for (let i = 1; i < n2; i++) {
            tail2.next = new Node(input_ar2[i]);
            tail2 = tail2.next;
        }
        let obj = new Solution();
        if (obj.areIdentical(head1, head2)) {
            console.log("true");
        } else {
            console.log("false");
        }
    }
}
// } Driver Code Ends


// User function Template for javascript

/**
 * @param {Node} head1
 * @param {Node} head2
 * @returns {boolean}
 */

/*
class Node{
    constructor(data){
        this.data = data;
        this.next = null;
    }
}
*/

class Solution {
    // Function to check whether two linked lists are identical or not.
    areIdentical(head1, head2) {
    //Your Code Start
              while (head1 != null && head2 != null) {
            if (head1.data != head2.data) {
                return false;
            }
            head1 = head1.next;
            head2 = head2.next;
        }
        return head1 == null && head2 == null;
    }
    //Your Code End
}

File: June/29-06-2024/Identical Linked Lists.c
//{ Driver Code Starts
// Initial Template for C

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node *next;
};


// } Driver Code Ends
// User function Template for C

// Function to check if 2 linked lists are identical
bool areIdentical(struct Node *head1, struct Node *head2) {
    //Your Code Start
        // Traverse both lists and compare data
    while (head1 != NULL && head2 != NULL) {
        if (head1->data != head2->data) {
            return false; // Data mismatch
        }
        head1 = head1->next;
        head2 = head2->next;
    }
    
    // Check if both lists have reached the end
    if (head1 == NULL && head2 == NULL) {
        return true; // Both lists are identical
    }
    
    return false; // Lists are of different lengths
//Your Code End
}

//{ Driver Code Starts.

int main() {
    int t;
    scanf("%d", &t);

    while (t--) {
        int n;
        scanf("%d", &n);
        int first;
        scanf("%d", &first);
        struct Node *head1;
        head1 = (struct Node *)malloc(sizeof(struct Node));
        head1->data = first;
        head1->next = NULL;
        struct Node *tail = head1;

        for (int i = 1; i < n; ++i) {
            int data;
            scanf("%d", &data);
            struct Node *temp;
            temp = (struct Node *)malloc(sizeof(struct Node));
            temp->data = data;
            temp->next = NULL;
            tail->next = temp;
            tail = tail->next;
        }

        scanf("%d", &n);
        scanf("%d", &first);
        struct Node *head2;
        head2 = (struct Node *)malloc(sizeof(struct Node));
        head2->data = first;
        head2->next = NULL;
        tail = head2;

        for (int i = 1; i < n; ++i) {
            int data;
            scanf("%d", &data);
            struct Node *temp;
            temp = (struct Node *)malloc(sizeof(struct Node));
            temp->data = data;
            temp->next = NULL;
            tail->next = temp;
            tail = tail->next;
        }
        areIdentical(head1, head2) ? printf("true\n") : printf("false\n");
    }

    return 0;
}

// } Driver Code Ends
